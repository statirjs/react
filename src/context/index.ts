import { createContext } from 'react';
import { Listener } from '@statirjs/core';
import * as S from '../typing/internal';

const context: S.ContextValue = {
  store: {
    state: {},
    dispatch: {},
    listeners: [],
    subscribe(listener: Listener) {
      this.listeners.push(listener);
    }
  }
};

export const Context = createContext(context);
