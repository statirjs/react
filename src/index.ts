export { Provider } from './provider';
export { Context } from './context';
export { useDispatch, useSelect } from './hooks';
