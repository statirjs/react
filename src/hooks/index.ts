import { useMemo, useContext } from 'react';
import { Context } from '../context';
import * as S from '../typing/internal';

export function useSelect<T>(selector: S.StateSelector<T>): T {
  const { store } = useContext(Context);
  const memoSelector = useMemo(() => selector, []);
  return memoSelector(store.state);
}

export function useDispatch<T extends S.Action>(
  selector: S.DispatchSelector<T>
): T {
  const { store } = useContext(Context);
  const memoSelector = useMemo(() => selector, []);
  return memoSelector(store.dispatch);
}
