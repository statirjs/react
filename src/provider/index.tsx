import React, { useState, useEffect, useMemo } from 'react';
import { Context } from '../context';
import * as S from '../typing/internal';

const RELOADER = {};

export function Provider(props: S.ProviderProps) {
  const { store, children = null } = props;

  const [reloader, setReloader] = useState(RELOADER);

  const value = useMemo(() => ({ store }), [reloader]);

  useEffect(() => {
    store.subscribe(() => {
      setReloader({ ...RELOADER });
    });
  }, []);

  return <Context.Provider value={value}>{children}</Context.Provider>;
}
