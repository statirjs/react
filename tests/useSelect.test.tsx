import React from 'react';
import renderer from 'react-test-renderer';
import { createForme, initStore } from '@statirjs/core';
import { Provider, useSelect } from '../src';

const counter = createForme(
  {
    count: 0
  },
  () => ({
    actions: {
      increment(state) {
        return {
          ...state,
          count: state.count + 1
        };
      }
    }
  })
);

const store = initStore({
  formes: {
    counter
  }
});

type RootState = typeof store.state;

function App() {
  const count = useSelect((rootState: RootState) => rootState.counter.count);

  return <span>{count}</span>;
}

describe('Test useSelect', () => {
  test('snapshot', () => {
    const component = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('value', () => {
    const component = renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );

    const tree = component.toJSON();

    expect(tree?.children?.[0]).toEqual('0');
  });
});
