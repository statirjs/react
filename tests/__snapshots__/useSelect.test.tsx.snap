// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test useSelect snapshot 1`] = `
<span>
  0
</span>
`;
